### ISSUE TITLE
[策略8] MACD多空策略 | MACD Long Short Strategy

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4426]
**标签**: quantitative-strategy, technical-analysis, macd, momentum
### 摘要
使用MACD指标的金叉死叉信号，结合成交量过滤进行多空交易。

### 算法描述
**核心原理：**
MACD（Moving Average Convergence Divergence）通过12日和26日指数移动平均线的差值计算MACD线，再与其9日信号线比较来判断趋势方向。金叉（MACD上穿信号线）做多，死叉（MACD下穿信号线）做空。

**核心公式：**
1. **MACD线计算：**
   ```
   MACD = EMA(close, 12) - EMA(close, 26)
   Signal_Line = EMA(MACD, 9)
   Histogram = MACD - Signal_Line
   ```

2. **标准化MACD（用于多标的比较）：**
   ```
   MACD_norm = (MACD - μ_MACD) / σ_MACD
   其中 μ=均值, σ=标准差, 计算窗口通常为60日
   ```

3. **趋势确认指标：**
   ```
   Trend_Score = w1*MACD_direction + w2*Price_position + w3*Volume_confirmation
   其中 w1+w2+w3=1, 权重可通过回测优化
   ```

4. **开仓信号：**
   ```
   Long_Signal = MACD > Signal_Line AND Histogram > 0 AND Volume > 20日均值*1.2
   Short_Signal = MACD < Signal_Line AND Histogram < 0 AND Volume > 20日均值*1.2
   ```

5. **止损止盈：**
   ```
   Stop_Loss = Entry_Price * (1 - 2 * ATR_20 / Entry_Price)
   Take_Profit = Entry_Price * (1 + 3 * ATR_20 / Entry_Price)
   ```

### Python代码示例
```python
import numpy as np
import pandas as pd
from dataclasses import dataclass
from typing import List, Dict, Optional, Tuple

@dataclass
class MACDConfig:
    fast_period: int = 12
    slow_period: int = 26
    signal_period: int = 9
    volume_ma_period: int = 20
    volume_threshold: float = 1.2

@dataclass
class MACDSignal:
    symbol: str
    direction: str  # 'long' or 'short'
    entry_price: float
    macd_value: float
    signal_line: float
    histogram: float
    volume_ratio: float
    confidence: float
    stop_loss: float
    take_profit: float

class MACDLongShortStrategy:
    def __init__(self, config: MACDConfig = None):
        self.config = config or MACDConfig()
        
    def calculate_ema(self, data: pd.Series, period: int) -> pd.Series:
        return data.ewm(span=period, adjust=False).mean()
    
    def calculate_macd(self, close: pd.Series) -> Tuple[pd.Series, pd.Series, pd.Series]:
        ema_fast = self.calculate_ema(close, self.config.fast_period)
        ema_slow = self.calculate_ema(close, self.config.slow_period)
        
        macd_line = ema_fast - ema_slow
        signal_line = self.calculate_ema(macd_line, self.config.signal_period)
        histogram = macd_line - signal_line
        
        return macd_line, signal_line, histogram
    
    def calculate_atr(self, high: pd.Series, low: pd.Series, 
                     close: pd.Series, period: int = 20) -> pd.Series:
        tr1 = high - low
        tr2 = abs(high - close.shift(1))
        tr3 = abs(low - close.shift(1))
        
        tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)
        atr = tr.rolling(period).mean()
        
        return atr
    
    def calculate_volume_ratio(self, volume: pd.Series) -> pd.Series:
        ma_volume = volume.rolling(self.config.volume_ma_period).mean()
        ratio = volume / ma_volume
        return ratio
    
    def calculate_normalized_macd(self, macd_series: pd.Series, 
                                  lookback: int = 60) -> pd.Series:
        rolling_mean = macd_series.rolling(lookback).mean()
        rolling_std = macd_series.rolling(lookback).std()
        return (macd_series - rolling_mean) / (rolling_std + 1e-10)
    
    def detect_crossover(self, macd: pd.Series, signal: pd.Series,
                        prev_macd: pd.Series, prev_signal: pd.Series) -> str:
        if prev_macd < prev_signal and macd > signal:
            return 'golden_cross'
        elif prev_macd > prev_signal and macd < signal:
            return 'dead_cross'
        return 'none'
    
    def calculate_trend_score(self, macd: float, signal: float,
                             histogram: float, price_position: float,
                             volume_ratio: float) -> float:
        macd_direction = 1 if histogram > 0 else -1
        
        price_pos_score = 1 if price_position > 0.5 else -1
        
        volume_score = 1 if volume_ratio > self.config.volume_threshold else 0
        
        w1, w2, w3 = 0.4, 0.35, 0.25
        score = w1 * macd_direction + w2 * price_pos_score + w3 * volume_score
        
        return np.clip(score, -1, 1)
    
    def generate_signals(self, data: pd.DataFrame, symbol: str) -> List[MACDSignal]:
        close = data['close']
        high = data.get('high', close)
        low = data.get('low', close)
        volume = data['volume']
        
        macd_line, signal_line, histogram = self.calculate_macd(close)
        atr = self.calculate_atr(high, low, close)
        volume_ratio = self.calculate_volume_ratio(volume)
        
        n = len(close)
        signals = []
        
        for i in range(1, n):
            if i < 60:
                continue
            
            macd_curr = macd_line.iloc[i]
            signal_curr = signal_line.iloc[i]
            hist_curr = histogram.iloc[i]
            hist_prev = histogram.iloc[i-1]
            
            macd_prev = macd_line.iloc[i-1]
            signal_prev = signal_line.iloc[i-1]
            
            crossover = self.detect_crossover(macd_curr, signal_curr, macd_prev, signal_prev)
            
            if crossover == 'none':
                continue
            
            vol_ratio = volume_ratio.iloc[i]
            atr_val = atr.iloc[i]
            entry_price = close.iloc[i]
            
            price_position = (close.iloc[i] - close.iloc[i-20:i].min()) / \
                           (close.iloc[i-20:i].max() - close.iloc[i-20:i].min() + 1e-10)
            
            trend_score = self.calculate_trend_score(
                macd_curr, signal_curr, hist_curr, price_position, vol_ratio
            )
            
            if crossover == 'golden_cross' and trend_score > 0.3:
                stop_loss = entry_price - 2 * atr_val
                take_profit = entry_price + 3 * atr_val
                
                signals.append(MACDSignal(
                    symbol=symbol,
                    direction='long',
                    entry_price=entry_price,
                    macd_value=macd_curr,
                    signal_line=signal_curr,
                    histogram=hist_curr,
                    volume_ratio=vol_ratio,
                    confidence=min(trend_score, 0.95),
                    stop_loss=stop_loss,
                    take_profit=take_profit
                ))
            
            elif crossover == 'dead_cross' and trend_score < -0.3:
                stop_loss = entry_price + 2 * atr_val
                take_profit = entry_price - 3 * atr_val
                
                signals.append(MACDSignal(
                    symbol=symbol,
                    direction='short',
                    entry_price=entry_price,
                    macd_value=macd_curr,
                    signal_line=signal_curr,
                    histogram=hist_curr,
                    volume_ratio=vol_ratio,
                    confidence=min(abs(trend_score), 0.95),
                    stop_loss=stop_loss,
                    take_profit=take_profit
                ))
        
        return signals
    
    def calculate_position_size(self, signal: MACDSignal,
                               portfolio_capital: float = 10000000,
                               risk_per_trade: float = 0.01) -> dict:
        risk_amount = portfolio_capital * risk_per_trade
        price_risk = abs(signal.entry_price - signal.stop_loss)
        
        if price_risk == 0:
            return {'shares': 0, 'position_value': 0}
        
        shares = int(risk_amount / price_risk)
        position_value = shares * signal.entry_price
        
        return {
            'shares': shares,
            'position_value': position_value,
            'risk_amount': risk_amount,
            'entry_price': signal.entry_price,
            'stop_loss': signal.stop_loss,
            'take_profit': signal.take_profit,
            'confidence': signal.confidence
        }

# 示例运行
if __name__ == '__main__':
    np.random.seed(42)
    n = 500
    prices = 100 * np.exp(np.cumsum(np.random.randn(n) * 0.02))
    volumes = np.random.randint(1000000, 10000000, n)
    
    data = pd.DataFrame({
        'close': prices,
        'high': prices * 1.02,
        'low': prices * 0.98,
        'volume': volumes
    })
    
    strategy = MACDLongShortStrategy()
    signals = strategy.generate_signals(data, 'BTCUSDT')
    
    print(f"Generated {len(signals)} signals")
    for sig in signals[:5]:
        pos = strategy.calculate_position_size(sig)
        print(f"{sig.direction.upper()}: Entry=${sig.entry_price:.2f}, "
              f"Conf={sig.confidence:.2%}, Stop=${sig.stop_loss:.2f}")
        print(f"  Position: {pos['shares']} shares = ${pos['position_value']:,.0f}")
